tidy3d/web/api/autograd/autograd.py

```diff
@@ -0,0 +1,698 @@
+# autograd wrapper for web functions
+
+import tidy3d as td
+from tidy3d.components.autograd import primitive, defvjp, AutogradFieldMap, get_static  # noqa: 401
+import typing
+import traceback
+
+import numpy as np
+from autograd.builtins import dict as dict_ag
+
+from ..webapi import run as run_webapi
+from ..asynchronous import run_async as run_async_webapi
+from ..asynchronous import DEFAULT_DATA_DIR
+from ..container import BatchData
+from ..tidy3d_stub import SimulationType, SimulationDataType
+
+from .utils import split_list, split_data_list, get_derivative_maps
+
+
+# keys for data into auxiliary dictionary
+AUX_KEY_SIM_DATA_ORIGINAL = "sim_data"
+AUX_KEY_SIM_DATA_FWD = "sim_data_fwd_adjoint"
+ISSUE_URL = (
+    "https://github.com/flexcompute/tidy3d/issues/new?"
+    "assignees=tylerflex&labels=adjoint&projects=&template=autograd_bug.md"
+)
+URL_LINK = f"[blue underline][link={ISSUE_URL}]'{ISSUE_URL}'[/link][/blue underline]"
+
+MAX_NUM_TRACED_STRUCTURES = 500
+
+
+def warn_autograd(fn_name: str, exc: Exception) -> str:
+    """Get warning message."""
+
+    exc_str = exc.__repr__()
+    traceback_str = "".join(traceback.format_tb(exc.__traceback__))
+
+    td.log.warning(
+        f"Autograd compatible '{fn_name}' failed, running original '{fn_name}'. "
+        "If you received this warning, please file an issue at the tidy3d front end with this "
+        f"message pasted in and we will investigate. \n\n "
+        f"{URL_LINK}.\n\n"
+        f"{exc_str} {traceback_str}.\n\n"
+    )
+
+
+def is_valid_for_autograd(simulation: td.Simulation) -> bool:
+    """Check whether a supplied simulation can use autograd run."""
+
+    # only support Simulations
+    if not isinstance(simulation, td.Simulation):
+        return False
+
+    # if no tracers just use regular web.run()
+    traced_fields = simulation.strip_traced_fields(
+        include_untraced_data_arrays=False, starting_path=("structures",)
+    )
+    if not traced_fields:
+        return False
+
+    # if too many structures, raise an error
+    structure_indices = [i for key, i, *_ in traced_fields.keys() if key == "structures"]
+    num_traced_structures = len(structure_indices)
+    if num_traced_structures > MAX_NUM_TRACED_STRUCTURES:
+        raise ValueError(
+            f"Autograd support is currently limited to {MAX_NUM_TRACED_STRUCTURES} structures with "
+            f"traced fields. Found {num_traced_structures} structures with traced fields."
+        )
+
+    return True
+
+
+def is_valid_for_autograd_async(simulations: dict[str, td.Simulation]) -> bool:
```

Comment:
I'd suggest to just return the result of those checks:
```python
def is_valid_for_autograd_async(simulations: dict[str, td.Simulation]) -> bool:
    return isinstance(simulations, dict) and all(
        is_valid_for_autograd(sim) for sim in simulations.values()
    )
```

Reply:
yea I refactored this earlier today and it does this and is a bit cleaner overall.